File: android/src/main/java/com/huaweiobs/core/ObsClientHolder.java
package com.huaweiobs.core;

import android.util.Log;

import com.obs.services.ObsClient;
import com.obs.services.ObsConfiguration;
import com.huaweiobs.utils.OBSException;

import java.util.HashMap;
import java.util.Map;

/**
 * OBS 客户端持有者
 * 封装华为云官方 ObsClient，管理配置和生命周期
 */
public class ObsClientHolder {
    private static final String TAG = "ObsClientHolder";
    private ObsClient obsClient;
    private Map<String, Object> currentConfig;

    /**
     * 创建 OBS 客户端
     */
    public synchronized void createClient(Map<String, Object> config) throws OBSException {
        closeClient();

        try {
            String endpoint = (String) config.get("endpoint");
            if (endpoint == null) {
                throw new IllegalArgumentException("Missing endpoint");
            }
            // setEndPoint 只接受主机名，去掉 http:// / https:// 前缀
            endpoint = endpoint.replaceFirst("^https?://", "");

            Log.i(TAG, "[Init] endpoint=" + endpoint
                    + "  bucket=" + config.get("bucket")
                    + "  socketTimeout=" + config.get("socketTimeout")
                    + "  connectionTimeout=" + config.get("connectionTimeout")
                    + "  maxRetry=" + config.get("maxErrorRetry"));

            String accessKeyId = (String) config.get("accessKeyId");
            if (accessKeyId == null) {
                throw new IllegalArgumentException("Missing accessKeyId");
            }

            String secretAccessKey = (String) config.get("secretAccessKey");
            if (secretAccessKey == null) {
                throw new IllegalArgumentException("Missing secretAccessKey");
            }

            ObsConfiguration obsConfiguration = new ObsConfiguration();
            obsConfiguration.setEndPoint(endpoint);
            
            Boolean isHttps = (Boolean) config.get("isHttps");
            obsConfiguration.setHttpsOnly(isHttps != null ? isHttps : true);
            
            // RN bridge 传来的数字类型为 Double，需通过 Number 中转再取 int
            Number socketTimeoutNum = (Number) config.get("socketTimeout");
            obsConfiguration.setSocketTimeout(socketTimeoutNum != null ? socketTimeoutNum.intValue() : 60000);
            
            Number connectionTimeoutNum = (Number) config.get("connectionTimeout");
            obsConfiguration.setConnectionTimeout(connectionTimeoutNum != null ? connectionTimeoutNum.intValue() : 60000);
            
            Number maxErrorRetryNum = (Number) config.get("maxErrorRetry");
            obsConfiguration.setMaxErrorRetry(maxErrorRetryNum != null ? maxErrorRetryNum.intValue() : 3);
            
            Boolean pathStyle = (Boolean) config.get("pathStyle");
            obsConfiguration.setPathStyle(pathStyle != null ? pathStyle : false);

            // 创建客户端
            // 注意：空字符串与 null 同等对待，避免用空 token 初始化导致认证失败
            String securityToken = (String) config.get("securityToken");
            if (securityToken != null && !securityToken.isEmpty()) {
                // STS 临时凭证（临时 AK/SK + Security Token）
                Log.i(TAG, "[Init] credential=STS  AK_prefix=" + accessKeyId.substring(0, Math.min(6, accessKeyId.length())));
                obsClient = new ObsClient(accessKeyId, secretAccessKey, securityToken, obsConfiguration);
            } else {
                // 永久 AK/SK
                Log.i(TAG, "[Init] credential=PERMANENT  AK_prefix=" + accessKeyId.substring(0, Math.min(6, accessKeyId.length())));
                obsClient = new ObsClient(accessKeyId, secretAccessKey, obsConfiguration);
            }

            currentConfig = new HashMap<>(config);
            Log.i(TAG, "[Init] OBS client created successfully");
        } catch (Exception e) {
            Log.e(TAG, "[Init] Failed to create OBS client: " + e.getMessage(), e);
            throw new OBSException("E_INIT_FAILED", "Failed to create OBS client: " + e.getMessage(), e);
        }
    }

    /**
     * 更新配置
     */
    public synchronized void updateConfig(Map<String, Object> newConfig) throws OBSException {
        Map<String, Object> mergedConfig = new HashMap<>();
        if (currentConfig != null) {
            mergedConfig.putAll(currentConfig);
        }
        mergedConfig.putAll(newConfig);
        createClient(mergedConfig);
    }

    /**
     * 获取客户端实例
     */
    public synchronized ObsClient getClient() throws OBSException {
        if (obsClient == null) {
            throw new OBSException("E_UNKNOWN", "OBS client not initialized", null);
        }
        return obsClient;
    }

    /**
     * 检查凭证是否过期
     */
    public boolean isCredentialsExpired() {
        if (currentConfig == null) {
            return false;
        }
        Number expiryTime = (Number) currentConfig.get("tokenExpiryTime");
        if (expiryTime == null) {
            return false;
        }
        return System.currentTimeMillis() > expiryTime.longValue();
    }

    /**
     * 获取当前配置
     */
    public Map<String, Object> getConfig() {
        return currentConfig != null ? new HashMap<>(currentConfig) : new HashMap<>();
    }

    /**
     * 关闭客户端
     */
    public synchronized void closeClient() {
        if (obsClient != null) {
            Log.i(TAG, "[Close] Closing OBS client");
        }
        try {
            if (obsClient != null) {
                obsClient.close();
            }
        } catch (Exception e) {
            Log.w(TAG, "[Close] Error while closing OBS client: " + e.getMessage());
        } finally {
            obsClient = null;
        }
    }
}

File: ios/Core/ObsClientHolder.m
//
//  ObsClientHolder.m
//  HuaweiObs
//
//  Created by react-native-huawei-obs
//

#import "ObsClientHolder.h"
#import <OBS/OBS.h>

#define OBSLog(tag, fmt, ...) NSLog(@"[%@] " fmt, tag, ##__VA_ARGS__)
static NSString *const TAG = @"ObsClientHolder";

@interface ObsClientHolder ()

@property (nonatomic, strong) OBSClient *obsClient;
@property (nonatomic, strong) NSString *endpoint;
@property (nonatomic, strong) NSString *bucket;
@property (nonatomic, strong) NSString *accessKeyId;
@property (nonatomic, strong) NSString *secretAccessKey;
@property (nonatomic, strong) NSString *securityToken;
@property (nonatomic, assign) NSTimeInterval tokenExpiryTime;
@property (nonatomic, strong) NSLock *lock;

@end

@implementation ObsClientHolder

- (instancetype)init {
    if (self = [super init]) {
        _lock = [[NSLock alloc] init];
    }
    return self;
}

- (void)createClientWithConfig:(NSDictionary *)config error:(NSError **)error {
    [self.lock lock];
    @try {
        // Parse required config
        NSString *endpoint = config[@"endpoint"];
        NSString *bucket = config[@"bucket"];
        NSString *accessKeyId = config[@"accessKeyId"];
        NSString *secretAccessKey = config[@"secretAccessKey"];
        
        if (!endpoint || !bucket) {
            if (error) {
                *error = [NSError errorWithDomain:@"ObsClientHolder" code:-1
                                         userInfo:@{NSLocalizedDescriptionKey: @"Missing required config: endpoint or bucket"}];
            }
            return;
        }
        
        OBSLog(TAG, @"[Init] endpoint=%@  bucket=%@  timeout=%@  maxRetries=%@",
               endpoint, bucket, config[@"timeout"], config[@"maxRetries"]);
        
        // Create credential provider
        OBSStaticCredentialProvider *credentialProvider = [[OBSStaticCredentialProvider alloc]
            initWithAccessKey:accessKeyId ?: @""
                    secretKey:secretAccessKey ?: @""];
        
        // Handle security token for STS
        NSString *securityToken = config[@"securityToken"];
        if (securityToken && securityToken.length > 0) {
            credentialProvider.securityToken = securityToken;
        }
        
        // Create service configuration
        NSURL *endpointURL = [NSURL URLWithString:endpoint];
        OBSServiceConfiguration *configuration = [[OBSServiceConfiguration alloc]
            initWithURL:endpointURL
            credentialProvider:credentialProvider];
        
        // Note: SDK 3.25.9 doesn't expose timeout and retry configuration properties
        // These are managed internally by the SDK
        
        // Create OBS client
        self.obsClient = [[OBSClient alloc] initWithConfiguration:configuration];
        
        // Store config
        self.endpoint = endpoint;
        self.bucket = bucket;
        self.accessKeyId = accessKeyId;
        self.secretAccessKey = secretAccessKey;
        self.securityToken = config[@"securityToken"];
        
        NSNumber *tokenExpiryTimeNum = config[@"tokenExpiryTime"];
        self.tokenExpiryTime = tokenExpiryTimeNum ? [tokenExpiryTimeNum doubleValue] : 0;
        
        NSString *credType = (self.securityToken && self.securityToken.length > 0) ? @"STS" : @"PERMANENT";
        OBSLog(TAG, @"[Init] credential=%@  AK_prefix=%@",
               credType,
               accessKeyId.length >= 6 ? [accessKeyId substringToIndex:6] : accessKeyId);
        OBSLog(TAG, @"[Init] OBS client created successfully");
        
    } @finally {
        [self.lock unlock];
    }
}

- (void)updateConfig:(NSDictionary *)config error:(NSError **)error {
    // 合并旧配置，避免部分更新时丢失其他配置
    NSMutableDictionary *mergedConfig = [NSMutableDictionary dictionary];
    if (self.endpoint) {
        [mergedConfig addEntriesFromDictionary:@{
            @"endpoint": self.endpoint ?: @"",
            @"bucket": self.bucket ?: @"",
            @"accessKeyId": self.accessKeyId ?: @"",
            @"secretAccessKey": self.secretAccessKey ?: @"",
        }];
        if (self.securityToken) mergedConfig[@"securityToken"] = self.securityToken;
        if (self.tokenExpiryTime > 0) mergedConfig[@"tokenExpiryTime"] = @(self.tokenExpiryTime);
    }
    [mergedConfig addEntriesFromDictionary:config];
    [self createClientWithConfig:mergedConfig error:error];
}

- (OBSClient *)getClientWithError:(NSError **)error {
    [self.lock lock];
    @try {
        if (!self.obsClient) {
            if (error) {
                *error = [NSError errorWithDomain:@"ObsClientHolder" code:-1
                                         userInfo:@{NSLocalizedDescriptionKey: @"OBS client not initialized"}];
            }
            return nil;
        }
        
        // Check token expiry
        if (self.tokenExpiryTime > 0) {
            NSTimeInterval currentTime = [[NSDate date] timeIntervalSince1970] * 1000;
            if (currentTime >= self.tokenExpiryTime) {
                if (error) {
                    *error = [NSError errorWithDomain:@"ObsClientHolder" code:-1
                                             userInfo:@{NSLocalizedDescriptionKey: @"Security token expired",
                                                       @"code": @"E_AUTH_EXPIRED"}];
                }
                return nil;
            }
        }
        
        return self.obsClient;
        
    } @finally {
        [self.lock unlock];
    }
}

- (BOOL)isCredentialsExpired {
    if (self.tokenExpiryTime == 0) {
        return NO;
    }
    NSTimeInterval currentTime = [[NSDate date] timeIntervalSince1970] * 1000;
    return currentTime >= self.tokenExpiryTime;
}

- (NSString *)getEndpoint {
    return self.endpoint;
}

- (NSDictionary *)getCurrentConfig {
    NSMutableDictionary *config = [NSMutableDictionary dictionary];
    if (self.endpoint) config[@"endpoint"] = self.endpoint;
    if (self.bucket) config[@"bucket"] = self.bucket;
    if (self.accessKeyId) config[@"accessKeyId"] = self.accessKeyId;
    if (self.secretAccessKey) config[@"secretAccessKey"] = self.secretAccessKey;
    if (self.securityToken) config[@"securityToken"] = self.securityToken;
    if (self.tokenExpiryTime > 0) config[@"tokenExpiryTime"] = @(self.tokenExpiryTime);
    return [config copy];
}

- (void)closeClient {
    [self.lock lock];
    if (self.obsClient) {
        OBSLog(TAG, @"[Close] Closing OBS client");
    }
    self.obsClient = nil;
    self.endpoint = nil;
    self.bucket = nil;
    self.accessKeyId = nil;
    self.secretAccessKey = nil;
    self.securityToken = nil;
    self.tokenExpiryTime = 0;
    [self.lock unlock];
}

@end
